maingate/src/range.rs

```diff
@@ -237,44 +279,43 @@ impl<F: FieldExt> RangeChip<F> {
             main_gate_config.d,
         );
 
-        macro_rules! meta_lookup {
-            ($prefix:literal, $column:expr, $table_config:expr) => {
-                meta.lookup(concat!($prefix, "_", stringify!($column)), |meta| {
-                    let exp = meta.query_advice($column, Rotation::cur());
-                    let s = meta.query_selector($table_config.selector);
-                    vec![(exp * s, $table_config.column)]
-                });
-            };
-        }
-
-        let mut composition_tables = BTreeMap::<usize, TableConfig>::new();
-        let mut overflow_tables = BTreeMap::<usize, TableConfig>::new();
+        let s_composition = meta.complex_selector();
+        let tag_composition = if composition_bit_lens.len() > 1 {
+            Some(meta.fixed_column())
+        } else {
+            None
+        };
+        lookup!(
+            "composition",
+            s_composition,
+            tag_composition,
+            composition_bit_lens,
+            [a, b, c, d]
+        );
 
-        for bit_len in composition_bit_lens.iter() {
-            let config = TableConfig {
-                selector: meta.complex_selector(),
-                column: meta.lookup_table_column(),
-            };
-            meta_lookup!("composition", a, config);
-            meta_lookup!("composition", b, config);
-            meta_lookup!("composition", c, config);
-            meta_lookup!("composition", d, config);
-            composition_tables.insert(*bit_len, config);
-        }
-        for bit_len in overflow_bit_lens.iter() {
-            let config = TableConfig {
-                selector: meta.complex_selector(),
-                column: meta.lookup_table_column(),
+        let (s_overflow, tag_overflow) = if has_overflow_limb {
```

Comment:
Maybe? Just a suggestion to save the creation of a variable that might be avoidable. I say it more for ease of read than performance of course.
```suggestion
        let (s_overflow, tag_overflow) = if !overflow_bit_lens.is_empty() {
```

Reply:
Updated in https://github.com/privacy-scaling-explorations/halo2wrong/pull/38/commits/0ca396e35d51807a87de72e776a20081540808d4